File: sdks/dotnet/src/Rrflow.Rrd.Client/OperationBinding.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Net.Http;

namespace Rrflow.Rrd;

internal sealed class OperationBinding
{
    private static readonly HashSet<string> ResourceKinds = new(StringComparer.Ordinal)
    {
        "organization", "estate", "project", "instance", "node", "shard", "collection",
        "table", "record", "transaction", "snapshot", "backup", "operation",
    };

    private OperationBinding(
        Endpoint endpoint,
        Uri target,
        Dictionary<string, object?>? requestContext,
        IReadOnlyList<ResourceSegment> resource,
        RequestOptions options)
    {
        Endpoint = endpoint;
        Target = target;
        RequestContext = requestContext;
        Resource = resource;
        Options = options;
    }

    internal Endpoint Endpoint { get; }
    internal Uri Target { get; }
    internal Dictionary<string, object?>? RequestContext { get; }
    internal IReadOnlyList<ResourceSegment> Resource { get; }
    private RequestOptions Options { get; }

    internal static OperationBinding Create(
        OperationId operation,
        RequestOptions options,
        EndpointResolver resolver)
    {
        Endpoint endpoint = EndpointCatalog.Get(operation);
        Dictionary<string, object?>? requestContext = MakeContext(endpoint, options);
        Uri target = resolver.Resolve(endpoint, options.PathParameters);
        IReadOnlyList<ResourceSegment> resource = endpoint.Method == "GET"
            ? Array.Empty<ResourceSegment>()
            : ValidateResource(
                options.Resource ?? DefaultResource(options.PathParameters, resolver.Instance));
        return new OperationBinding(endpoint, target, requestContext, resource, options);
    }

    internal void Authenticate(HttpRequestMessage request)
    {
        if (Endpoint.Authentication == Authentication.ApiKey)
        {
            if (Options.ApiKey is null || string.IsNullOrEmpty(Options.ApiKey.Credential))
            {
                throw new RrdClientException($"{Endpoint.WireName} requires API-key authentication");
            }
            request.Headers.Add(
                "X-RRD-Principal",
                EndpointResolver.CanonicalIdentifier(Options.ApiKey.PrincipalId, "principal"));
            request.Headers.TryAddWithoutValidation(
                "Authorization",
                "ApiKey " + Options.ApiKey.Credential);
        }
        else if (Endpoint.Authentication == Authentication.SessionBearer)
        {
            if (Options.Session is null
                || Options.Session.Lease.SessionId.Length == 0
                || Options.Session.Lease.Token.Length == 0)
            {
                throw new RrdClientException($"{Endpoint.WireName} requires a valid session");
            }
            request.Headers.Add("X-RRD-Session", Options.Session.Lease.SessionId);
            request.Headers.TryAddWithoutValidation(
                "Authorization",
                "Bearer " + Options.Session.Lease.Token);
        }
    }

    private static Dictionary<string, object?>? MakeContext(
        Endpoint endpoint,
        RequestOptions options)
    {
        if (endpoint.Method == "GET")
        {
            return null;
        }
        Dictionary<string, object?> context = new()
        {
            ["request_id"] = EndpointResolver.CorrelationIdentifier(
                options.RequestId,
                "request ID"),
            ["operation_id"] = EndpointResolver.CorrelationIdentifier(
                options.OperationId,
                "operation ID"),
        };
        if (options.IdempotencyKey is not null)
        {
            context["idempotency_key"] = EndpointResolver.CorrelationIdentifier(
                options.IdempotencyKey,
                "idempotency key");
        }
        else if (endpoint.Mutation)
        {
            throw new RrdClientException("mutating requests require an idempotency key");
        }
        if (options.Deadline is not null)
        {
            long milliseconds = options.Deadline.Value.ToUnixTimeMilliseconds();
            if (milliseconds <= 0)
            {
                throw new RrdClientException("deadline must be a positive Unix millisecond instant");
            }
            context["deadline_unix_ms"] = milliseconds;
        }
        return context;
    }

    private static IReadOnlyList<ResourceSegment> DefaultResource(
        IReadOnlyDictionary<string, string> parameters,
        string instance)
    {
        List<ResourceSegment> result = new();
        if (parameters.TryGetValue("estate", out string? estate))
        {
            result.Add(new ResourceSegment(
                "estate",
                EndpointResolver.CanonicalIdentifier(estate, "estate")));
        }
        result.Add(new ResourceSegment("instance", instance));
        return result;
    }

    private static IReadOnlyList<ResourceSegment> ValidateResource(
        IReadOnlyList<ResourceSegment> segments)
    {
        if (segments.Count is < 1 or > 16)
        {
            throw new RrdClientException("resource paths must contain 1..=16 segments");
        }
        HashSet<string> kinds = new(StringComparer.Ordinal);
        return segments.Select(segment =>
        {
            if (!ResourceKinds.Contains(segment.Kind) || !kinds.Add(segment.Kind))
            {
                throw new RrdClientException("resource kind is unknown or repeated");
            }
            return new ResourceSegment(
                segment.Kind,
                EndpointResolver.CanonicalIdentifier(segment.Id, "resource"));
        }).ToArray();
    }
}

File: sdks/java/src/main/java/io/rrflow/rrd/OperationBinding.java
package io.rrflow.rrd;

import java.net.http.HttpRequest;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;
import java.util.regex.Pattern;

/** Current common operation binding; H-04 owns generated concrete bindings. */
final class OperationBinding {
    private static final Pattern CORRELATION = Pattern.compile("^[A-Za-z0-9._:-]+$");
    private static final Pattern CANONICAL = Pattern.compile("^[a-z0-9][a-z0-9._-]*$");
    private static final Set<String> RESOURCE_KINDS = Set.of(
            "organization",
            "estate",
            "project",
            "instance",
            "node",
            "shard",
            "collection",
            "table",
            "record",
            "transaction",
            "snapshot",
            "backup",
            "operation");

    private final String instance;

    OperationBinding(String instance) {
        this.instance = instance;
    }

    Map<String, Object> requestContext(OperationId operation, RequestOptions options) {
        if (operation.method().equals("GET")) {
            return null;
        }
        Map<String, Object> context = new HashMap<>();
        context.put("request_id", correlation(options.requestId(), "request ID"));
        context.put("operation_id", correlation(options.operationId(), "operation ID"));
        if (options.idempotencyKey() != null) {
            context.put(
                    "idempotency_key",
                    correlation(options.idempotencyKey(), "idempotency key"));
        } else if (operation.mutation()) {
            throw new RrdClientException("mutating requests require an idempotency key");
        }
        if (options.deadline() != null) {
            long milliseconds = options.deadline().toEpochMilli();
            if (milliseconds <= 0) {
                throw new RrdClientException(
                        "deadline must be a positive Unix millisecond instant");
            }
            context.put("deadline_unix_ms", milliseconds);
        }
        return context;
    }

    void authenticate(
            HttpRequest.Builder builder, OperationId operation, RequestOptions options) {
        switch (operation.authentication()) {
            case PUBLIC -> { }
            case API_KEY -> {
                RequestOptions.ApiKey apiKey = options.apiKey();
                if (apiKey == null
                        || apiKey.credential() == null
                        || apiKey.credential().isEmpty()) {
                    throw new RrdClientException(
                            operation.wireName() + " requires API-key authentication");
                }
                builder.header(
                        "X-RRD-Principal", canonical(apiKey.principalId(), "principal"));
                builder.header("Authorization", "ApiKey " + apiKey.credential());
            }
            case SESSION_BEARER -> {
                Session session = options.session();
                if (session == null
                        || session.lease().sessionId().isEmpty()
                        || session.lease().token().isEmpty()) {
                    throw new RrdClientException(
                            operation.wireName() + " requires a valid session");
                }
                builder.header("X-RRD-Session", session.lease().sessionId());
                builder.header("Authorization", "Bearer " + session.lease().token());
            }
        }
    }

    List<ResourceSegment> resource(RequestOptions options) {
        List<ResourceSegment> resource = options.resource();
        if (resource == null) {
            resource = defaultResource(options.pathParameters());
        }
        return validateResource(resource);
    }

    private List<ResourceSegment> defaultResource(Map<String, String> parameters) {
        List<ResourceSegment> result = new ArrayList<>();
        String estate = parameters.get("estate");
        if (estate != null) {
            result.add(new ResourceSegment("estate", canonical(estate, "estate")));
        }
        result.add(new ResourceSegment("instance", instance));
        return result;
    }

    private static List<ResourceSegment> validateResource(List<ResourceSegment> segments) {
        if (segments.isEmpty() || segments.size() > 16) {
            throw new RrdClientException("resource paths must contain 1..=16 segments");
        }
        Set<String> kinds = new HashSet<>();
        List<ResourceSegment> result = new ArrayList<>();
        for (ResourceSegment segment : segments) {
            if (!RESOURCE_KINDS.contains(segment.kind()) || !kinds.add(segment.kind())) {
                throw new RrdClientException("resource kind is unknown or repeated");
            }
            result.add(new ResourceSegment(segment.kind(), canonical(segment.id(), "resource")));
        }
        return List.copyOf(result);
    }

    static String correlation(String value, String label) {
        if (value == null || value.length() > 128 || !CORRELATION.matcher(value).matches()) {
            throw new RrdClientException(label + " is not a canonical RRD correlation ID");
        }
        return value;
    }

    static String canonical(String value, String label) {
        if (value == null || value.length() > 128 || !CANONICAL.matcher(value).matches()) {
            throw new RrdClientException(label + " is not a canonical RRD identifier");
        }
        return value;
    }
}
